File: java/connect/src/main/java/Connect.java
import java.util.*;

class Connect {
    private char[][] board;
    private int rows;
    private int cols;

    public Connect(String[] boardInput) {
        // Parse the board representation
        rows = boardInput.length;
        
        // Find the maximum length of a row after removing spaces
        cols = 0;
        for (String row : boardInput) {
            // Count non-space characters
            int nonSpaceCount = 0;
            for (char c : row.toCharArray()) {
                if (c != ' ') {
                    nonSpaceCount++;
                }
            }
            cols = Math.max(cols, nonSpaceCount);
        }
        
        // Initialize the board
        board = new char[rows][cols];
        
        // Fill the board with the input
        for (int i = 0; i < rows; i++) {
            String row = boardInput[i];
            int col = 0;
            
            for (int j = 0; j < row.length(); j++) {
                char c = row.charAt(j);
                if (c != ' ') {
                    board[i][col] = c;
                    col++;
                }
            }
        }
    }

    public Winner computeWinner() {
        // Check if player X has won (connected left to right)
        if (hasPlayerXWon()) {
            return Winner.PLAYER_X;
        }
        
        // Check if player O has won (connected top to bottom)
        if (hasPlayerOWon()) {
            return Winner.PLAYER_O;
        }
        
        // No winner
        return Winner.NONE;
    }
    
    private boolean hasPlayerXWon() {
        // Check if player X has connected left to right
        boolean[][] visited = new boolean[rows][cols];
        
        // Start from each cell in the leftmost column
        for (int i = 0; i < rows; i++) {
            if (board[i][0] == 'X') {
                if (dfsForX(i, 0, visited)) {
                    return true;
                }
            }
        }
        
        return false;
    }
    
    private boolean hasPlayerOWon() {
        // Check if player O has connected top to bottom
        boolean[][] visited = new boolean[rows][cols];
        
        // Start from each cell in the top row
        for (int j = 0; j < cols; j++) {
            if (board[0][j] == 'O') {
                if (dfsForO(0, j, visited)) {
                    return true;
                }
            }
        }
        
        return false;
    }
    
    private boolean dfsForX(int row, int col, boolean[][] visited) {
        // If we've already visited this cell, return false
        if (visited[row][col]) {
            return false;
        }
        
        // Mark current cell as visited
        visited[row][col] = true;
        
        // If we reached the rightmost column, player X has won
        if (col == cols - 1) {
            return true;
        }
        
        // Check all adjacent cells
        for (int[] neighbor : getNeighbors(row, col)) {
            int newRow = neighbor[0];
            int newCol = neighbor[1];
            
            if (isValidPosition(newRow, newCol) && board[newRow][newCol] == 'X') {
                if (dfsForX(newRow, newCol, visited)) {
                    return true;
                }
            }
        }
        
        return false;
    }
    
    private boolean dfsForO(int row, int col, boolean[][] visited) {
        // If we've already visited this cell, return false
        if (visited[row][col]) {
            return false;
        }
        
        // Mark current cell as visited
        visited[row][col] = true;
        
        // If we reached the bottom row, player O has won
        if (row == rows - 1) {
            return true;
        }
        
        // Check all adjacent cells
        for (int[] neighbor : getNeighbors(row, col)) {
            int newRow = neighbor[0];
            int newCol = neighbor[1];
            
            if (isValidPosition(newRow, newCol) && board[newRow][newCol] == 'O') {
                if (dfsForO(newRow, newCol, visited)) {
                    return true;
                }
            }
        }
        
        return false;
    }
    
    private List<int[]> getNeighbors(int row, int col) {
        List<int[]> neighbors = new ArrayList<>();
        
        // In a hexagonal grid, each cell has 6 neighbors
        // The directions depend on the position in the grid
        
        // Horizontal neighbors (always present)
        if (col > 0) neighbors.add(new int[]{row, col - 1}); // Left
        if (col < cols - 1) neighbors.add(new int[]{row, col + 1}); // Right
        
        // Upper neighbors
        if (row > 0) {
            neighbors.add(new int[]{row - 1, col}); // Upper
            if (col < cols - 1) neighbors.add(new int[]{row - 1, col + 1}); // Upper-right
        }
        
        // Lower neighbors
        if (row < rows - 1) {
            neighbors.add(new int[]{row + 1, col}); // Lower
            if (col > 0) neighbors.add(new int[]{row + 1, col - 1}); // Lower-left
        }
        
        return neighbors;
    }
    
    private boolean isValidPosition(int row, int col) {
        return row >= 0 && row < rows && col >= 0 && col < cols;
    }
}

File: go/connect/connect.go
package connect

// ResultOf determines the winner of a Hex game.
// Player X wins by connecting left to right.
// Player O wins by connecting top to bottom.
// Returns "X" if X wins, "O" if O wins, or "" if there's no winner.
func ResultOf(lines []string) (string, error) {
	// Create a 2D board representation
	board := make([][]rune, len(lines))
	for i, line := range lines {
		board[i] = []rune(line)
	}

	// Check if player X has a path from left to right
	if hasWinningPath(board, 'X') {
		return "X", nil
	}

	// Check if player O has a path from top to bottom
	if hasWinningPath(board, 'O') {
		return "O", nil
	}

	// No winner
	return "", nil
}

// hasWinningPath checks if the specified player has a winning path
func hasWinningPath(board [][]rune, player rune) bool {
	rows := len(board)
	if rows == 0 {
		return false
	}

	// For a 1x1 board
	if rows == 1 && len(board[0]) == 1 {
		return board[0][0] == player
	}

	// Create a visited array
	visited := make([][]bool, rows)
	for i := range visited {
		visited[i] = make([]bool, len(board[i]))
	}

	// For player X, check if there's a path from left to right
	if player == 'X' {
		// Find starting points on the left edge
		for i := 0; i < rows; i++ {
			if i < len(board) && len(board[i]) > 0 && board[i][0] == 'X' {
				// Start DFS from this point
				if dfsX(board, visited, i, 0) {
					return true
				}
			}
		}
	}

	// For player O, check if there's a path from top to bottom
	if player == 'O' {
		// Reset visited array
		for i := range visited {
			for j := range visited[i] {
				visited[i][j] = false
			}
		}

		// Find starting points on the top edge
		if len(board[0]) > 0 {
			for j := 0; j < len(board[0]); j++ {
				if board[0][j] == 'O' {
					// Start DFS from this point
					if dfsO(board, visited, 0, j) {
						return true
					}
				}
			}
		}
	}

	return false
}

// dfsX performs depth-first search for player X (left to right)
func dfsX(board [][]rune, visited [][]bool, row, col int) bool {
	// Check if we've reached the right edge
	if col == len(board[row])-1 {
		return true
	}

	// Mark current cell as visited
	visited[row][col] = true

	// Define possible directions (including hexagonal adjacency)
	// In a hex grid, the adjacent cells depend on the row's parity
	directions := [][]int{
		{-1, 0}, {-1, 1}, // Up-left, Up-right
		{0, -1}, {0, 1},  // Left, Right
		{1, -1}, {1, 0},  // Down-left, Down-right
	}

	// Check all adjacent cells
	for _, dir := range directions {
		newRow, newCol := row+dir[0], col+dir[1]

		// Check if the new position is valid
		if newRow >= 0 && newRow < len(board) && 
		   newCol >= 0 && newCol < len(board[newRow]) && 
		   board[newRow][newCol] == 'X' && 
		   !visited[newRow][newCol] {
			// Continue DFS from this new position
			if dfsX(board, visited, newRow, newCol) {
				return true
			}
		}
	}

	return false
}

// dfsO performs depth-first search for player O (top to bottom)
func dfsO(board [][]rune, visited [][]bool, row, col int) bool {
	// Check if we've reached the bottom edge
	if row == len(board)-1 {
		return true
	}

	// Mark current cell as visited
	visited[row][col] = true

	// Define possible directions (including hexagonal adjacency)
	directions := [][]int{
		{-1, 0}, {-1, 1}, // Up-left, Up-right
		{0, -1}, {0, 1},  // Left, Right
		{1, -1}, {1, 0},  // Down-left, Down-right
	}

	// Check all adjacent cells
	for _, dir := range directions {
		newRow, newCol := row+dir[0], col+dir[1]

		// Check if the new position is valid
		if newRow >= 0 && newRow < len(board) && 
		   newCol >= 0 && newCol < len(board[newRow]) && 
		   board[newRow][newCol] == 'O' && 
		   !visited[newRow][newCol] {
			// Continue DFS from this new position
			if dfsO(board, visited, newRow, newCol) {
				return true
			}
		}
	}

	return false
}
